### ISSUE TITLE
Gradient accumulation broken in PP

### ISSUE THREAD

**Issue by jdinalt** [pytorch/torchtitan#1733]
### Bug description

Using gradient accumulation is incompatible with PipleineSchedule(..., scale_grads=True) option, which defaults to True.

When this option is set, at each step, all gradients are scaled by the micro-batch size. This works fine for a single gradient accumulation step, but when using multiple steps, this will rescale the total gradient by this factor, not just at the end of gradient accumulation.

The result is that the accumulated gradient is an exponential moving average, rather than a sum. Overall, the resulting gradients are much smaller than they should be and using gradient accumulation with PP is not equivalent to using it without PP -- the loss curves diverge substantially, as well as the gradient-norms are way off.

A secondary consequence is that at every step, it divides the gradients by n_microbatches, which is computationally expensive when applied to a large model.

I identified the same issue in my own pipeline trainer implementation a week or two ago. When checking how Torch Titan addressed the issue, I discovered that Titan probably has the same bug.

I had the time to confirm the presence of the issue today and have submitted https://github.com/pytorch/torchtitan/pull/1732 to resolve the issue.

### Versions

torch 2.10.0.dev20250915+cu126

For anyone who may be interested, I have added support for Torch Titan to my configuration framework, which is what I used for reproducing the issue.

https://github.com/jdinalt/forgather/tree/main/examples/torchtitan

**Comment by fegin** [comment]
Since PP already divides the batch size to micro batch size, I'm curious that is it common to use gradient accumulation step with PP?

**Comment by jdinalt** [comment]
It seems that dividing the batch into microbatches is orthogonal to using gradient accumulation?  The former is an implementation detail of pipeline parallelism, which is needed to keep all of the GPUs busy, while that latter is used to increase the effective batch size.

That is, if I don't have enough GPUs/memory to achieve the target batch size , I can use gradient accumulation to increase the effective batch size.

Perhaps I'm not understanding the question?

**Comment by fegin** [comment]
My interpretation of gradient accumulation is that given an effective batch size, we don't have enough GPUs/GPU memory to train, so we have to divide the batch size for each step and accumulate the gradients across steps. It's orthogonal to the microbatch but both of the techniques result in splitting batch size, and we couldn't split local batch size to be smaller than 1, hence my question.

**Comment by jdinalt** [comment]
The number of micro-batches does not change the memory requirements, for any given local batch size.

Consider a pipeline parallel configuration with two GPUs and a local batch size of 32.  A single batch of 32, two batches of 16, and 4 batches of 8, all have the same memory requirements.

## Naive Pipeline Schedule 1

Consider the case where we feed the whole batch of 32 into the pipeline at once .
- local_batch_size = 32
- n_microbatches = 1
-  microbatch_size = 32

1. 32 examples are passed to stage 1, which produces the input for stage 2.
2. 32 examples are fed into the 2nd stage., while the 1st stage is now idle.
3. The examples make it to the loss function at the end of the 2nd stage. Gradients are computed and sent back to the first stage. Stage 1 is still idle.
4. The gradients from the 2nd stage are forwarded to the first stage. Now, the 2nd stage is idle, while the 1st computes gradients.
5. All gradients have been computed
6. The optimizers applies all of the gradients in parallel on both stages and gradients are zero'd.

This "naive" pipeline approach highlights why we use micro-batches in the first place. We are trying to avoid letting our hardware sit idle.

In the above scenario, each GPU needs to have sufficient memory to hold all the activations generated by a batch size of 32. While we could introduce gradient checkpointing, the memory requirements are still directly proportional to the local batch size, which are invariant WRT to n_microbatches.

## Less-Bad Pipeline Example

Now, let's try something different. Instead of feeding all 32 examples through the pipeline at once, let's split this into two microbatches of 16.

- local_batch_size = 32
- n_microbatches = 2
-  microbatch_size = 16

1. Batch 1, of 16 examples, is fed into the 1st stage and the resulting activations are forwarded to the 2nd stage.
2. Batch 2  is fed into the first stage. This is concurrent with  stage two performing the forward pass on batch 1.
3. Now, the first stage has no further inputs, so is idle. The second stage now has two things it can work on. Either, the forward pass on the 2nd batch of the backward pass on the 1st batch. As we want to give stage 1 something to do, so we process the gradients first and pass them to the first stage.
4. Stage 1 begins processing the gradients from the 1st input batch. At the same time, the 2nd stage performs the forward pass on batch 2.
5. Stage 1 sits idle, while stage 2 computes the gradients from batch 2.
6. Stage 1 processes the gradients batch 2, while the 2nd stage is idle.
7. The optimizers applies all of the gradients in parallel on both stages and gradients are zero'd.

Still far from optimal, but is "less-bad."

In this second example, both stages have two separate gradient computation steps, and these are technically "accumulated" over those two steps.

The important thing to note is that stage 1 needs to have sufficient memory to buffer the activations from both the 1st and 2nd batch concurrently. You have not saved any activation memory by splitting the large batch into two micro-batches. In both the first and second pipeline scenario, the effective batch size is 32.

## Consequence

Assuming that this is the largest batch size that you have memory for, this is also the largest possible batch size you can train with.

But what if you need to train with larger batches? Let's say you really want a batch size of 320 for optimal performance. You can change the microbatch size to any combination of: 1x32, 2x16, 4x8, 16x2, or 32x1; your total effective batch size is still 32. 

That's where gradient-accumulation comes in. In the previous examples, we apply the optimizer step after all microbatches have completed, then clear the gradients for the next pass. What we can do instead is skip optimization and zero_grad and run 9 more passes through the pipeline.

This will cause the gradients to accumulate across all 320 examples. Then we apply our optimizer and reset the gradients. Bingo! This produces exactly where we were looking for.

While microbatches do accumulate gradients internally, it's for a completely different purpose than the technique known as "gradient accumulation." The pipeline does this to allow greater concurrency of processing than could be otherwise achieved. It is not intended to increase the effective batch size, nor can it be made to do so.

## The Problem

This is where this bug that I identified comes in. By default, the pipeline rescales the gradients by the number of microbatches at the end of each step. This keeps the gradients equivalent to what they would have been, had the whole batch been passed through at once -- it's computing the mean gradients for a batch size of 32.

 There is a problem when you combine this approach with gradient accumulation. The gradients are not just summed, which is what "gradient accumulation" is expected to do. They are summed and divided at each step.

Say, you are using 10 gradient accumulation steps. The first step should be equally weighted with the the other 9, having contributed 10% of the final gradients. Let's say there are four microbatches. At each step, the resulting gradients are summed and divided by 4. This is applied 10 times to the first gradient, thus it now has a weight of 1/4^10, 0.0001%, rather than 10%! This is an exponential moving average, rather than just an average.

I'll see if I can upload some Tensorboard plots of what happens when you don't address this problem and what happens when my changes are applied.

**Comment by jdinalt** [comment]
<img width="3730" height="1356" alt="Image" src="https://github.com/user-attachments/assets/4deaf0d7-7ac0-4e00-ace4-19812b67eeb9" />

In this Tensorboard screen-shot, it's pretty easy to see the issue. In all cases, the local batch size is 32 and the global batch size is 320 (10 gradient accumulation steps).

- control (orange) : This is the baseline, using a single GPU and nothing fancy.
- PP (red): This is a PP configuration with 2 GPUs and 2 stages, "1F1B." This does not have my changes.
- PP (blue): This is the same as above, but with my changes applied.
- TP (grey): Tensor Parallel, 2 GPUs.

Notice how low the grad_norm for "PP (red)." This is a consequence of the "accumulate and rescale logic," built into the pipeline scheduler.

Also note how the loss for "PP (red)" significantly diverges from the control, TP, and PP (blue), which are all nearly identical, as they should be. PP now matches the other plots, as it should.

I have traces for the other combinations on the list as well. They all match each other, but different a little for the selected cluster. It seems that anything involving data-parallel may have some going on, but it's not related to this issue and they are still pretty close to the displayed traces. 

The tools for reproducing this experiment are linked in my first message.

**Comment by tianyu-l** [comment]
Thank you for reporting the bug and proposing a fix! This is very valuable. In general I agree with the issue report, and proposed some changes to the fix.

> Since PP already divides the batch size to micro batch size, I'm curious that is it common to use gradient accumulation step with PP?

To address @fegin 's question more directly:
One might ask: Although grad accumulation helps reduce local batch size to fit a large global batch in memory, PP could already make microbatch size as small as 1. If one uses PP, wouldn't they always be able to use a large local batch size (without gradient accumulation) and microbatch size 1?

I think @jdinalt answers it:
> The important thing to note is that stage 1 needs to have sufficient memory to buffer the activations from both the 1st and 2nd batch concurrently. You have not saved any activation memory by splitting the large batch into two micro-batches.

In general (for most PP schedules?), the more microbatches we have, the more activation memory we need to keep for the backward. Even if we can use activation checkpointing, the checkpointed activation also grows linearly with the number of microbatches. So this scaling has a limit, which requires gradient accumulation to help scale further.

In their nature, PP is still doing computation "in parallel" (at the cost of saving activations in memory), but gradient accumulation is "sequential" (only keeping the gradient).

cc: @H-Huang as torchtitan probably will only use `scale_grads=False` in `_PipleineSchedule` for this reason.

**Comment by fegin** [comment]
Thanks for the detail writing. I didn't express myself clearly and understand that PP doesn't save the activation memory. I was just wondering how many use cases in reality people use this combination. But it is clearly a bug after reading your description. So we have to fix it.

**Comment by jdinalt** [comment]
@fegin 
> Thanks for the detail writing....

No problem. It was just not clear what you were asking.

While I don't have any hard data, my impression is that the usage of this feature is pretty common.

From personal experience, it's especially useful when training a model at the boundary of what you can fit into memory. In this case, one may reduce the local batch size to the absolute minimum required for the pipeline type, increase the context to as large as you can get a way with, and make up for the small batch size with gradient accumulation.

**Comment by jdinalt** [comment]
Where there's one missed copy of modified code...

https://github.com/pytorch/torchtitan/blob/0f34257d91da9db5de2ef654cc7cd2310dfa24c7/torchtitan/experiments/deepseek_v3/train_ds_dev.py#L129

I believe that's the last of them?
